File: benchmark/runners/python/infermap_bench/report.py
"""Build and validate report.json from case results."""
from __future__ import annotations

import json
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Callable, TypedDict

import jsonschema


class MetricSet(TypedDict):
    """Aggregated scorecard metrics for a slice of results."""
    f1: float
    top1: float
    mrr: float
    ece: float
    n: int


class Report(TypedDict):
    """Shape of a single-language benchmark report.json."""
    version: int
    language: str
    infermap_version: str
    runner_version: str
    ran_at: str
    duration_seconds: float
    scorecard: dict[str, Any]
    per_case: list[dict[str, Any]]
    failed_cases: list[str]

from . import REPORT_VERSION
from .metrics import (
    Prediction,
    expected_calibration_error,
    macro_mean,
    micro_f1,
)


@dataclass(frozen=True)
class CaseResult:
    case_id: str
    category: str
    subcategory: str
    difficulty: str
    tags: list[str]
    top1: float
    f1: float
    mrr: float
    tp: int
    fp: int
    fn: int
    predictions: list[Prediction]
    failed: bool = False
    failure_reason: str | None = None


def _round(n: float, digits: int = 6) -> float:
    return round(n, digits)


def _metric_set(results: list[CaseResult]) -> dict:
    counts = [(r.tp, r.fp, r.fn) for r in results]
    all_predictions: list[Prediction] = []
    for r in results:
        all_predictions.extend(r.predictions)
    return {
        "f1": _round(micro_f1(counts)),
        "top1": _round(macro_mean([r.top1 for r in results])),
        "mrr": _round(macro_mean([r.mrr for r in results])),
        "ece": _round(expected_calibration_error(all_predictions)),
        "n": len(results),
    }


def _by_single_key(
    results: list[CaseResult],
    key: Callable[[CaseResult], str],
) -> dict:
    buckets: dict[str, list[CaseResult]] = {}
    for r in results:
        buckets.setdefault(key(r), []).append(r)
    return {k: _metric_set(buckets[k]) for k in sorted(buckets)}


def _by_tag(results: list[CaseResult]) -> dict:
    buckets: dict[str, list[CaseResult]] = {}
    for r in results:
        for tag in r.tags:
            buckets.setdefault(tag, []).append(r)
    return {k: _metric_set(buckets[k]) for k in sorted(buckets)}


def build_report(
    results: list[CaseResult],
    *,
    language: str,
    infermap_version: str,
    runner_version: str,
    duration_seconds: float,
) -> Report:
    """Assemble a report.json dict from a list of per-case results.

    Aggregates metrics into overall, by_difficulty, by_category, and by_tag
    slices, and emits one per_case entry per result. The return value conforms
    to the `Report` TypedDict and the committed report.schema.json.
    """
    ran_at = datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")
    return {
        "version": REPORT_VERSION,
        "language": language,
        "infermap_version": infermap_version,
        "runner_version": runner_version,
        "ran_at": ran_at,
        "duration_seconds": duration_seconds,
        "scorecard": {
            "overall": _metric_set(results),
            "by_difficulty": _by_single_key(results, lambda r: r.difficulty),
            "by_category": _by_single_key(results, lambda r: r.category),
            "by_tag": _by_tag(results),
        },
        "per_case": [
            {
                "id": r.case_id,
                "f1": _round(r.f1),
                "top1": _round(r.top1),
                "mrr": _round(r.mrr),
                "expected_n": r.tp + r.fn,
                "predicted_n": r.tp + r.fp,
                "true_positives": r.tp,
                "false_positives": r.fp,
                "false_negatives": r.fn,
                "failed": r.failed,
                "failure_reason": r.failure_reason,
            }
            for r in results
        ],
        "failed_cases": [r.case_id for r in results if r.failed],
    }


def validate_report(report: dict, schema_path: Path | str) -> None:
    """Validate a report dict against report.schema.json. Raises on mismatch."""
    schema = json.loads(Path(schema_path).read_text(encoding="utf-8"))
    jsonschema.validate(report, schema)


def write_report(
    report: dict, output_path: Path | str, schema_path: Path | str
) -> None:
    """Validate a report and atomically write it as pretty-printed JSON."""
    validate_report(report, schema_path)
    Path(output_path).write_text(
        json.dumps(report, indent=2, sort_keys=False) + "\n",
        encoding="utf-8",
    )

File: benchmark/runners/ts/src/report.ts
// Build and validate report.json from case results.
// Mirrors infermap_bench/report.py.

import { readFileSync, writeFileSync } from "node:fs";
import Ajv2020 from "ajv/dist/2020.js";
import addFormats from "ajv-formats";
import {
  expectedCalibrationError,
  macroMean,
  microF1,
  type Prediction,
} from "./metrics.js";
import { REPORT_VERSION } from "./index.js";

export interface CaseResult {
  caseId: string;
  category: string;
  subcategory: string;
  difficulty: string;
  tags: string[];
  top1: number;
  f1: number;
  mrr: number;
  tp: number;
  fp: number;
  fn: number;
  predictions: Prediction[];
  failed: boolean;
  failureReason: string | null;
}

interface MetricSet {
  f1: number;
  top1: number;
  mrr: number;
  ece: number;
  n: number;
}

export interface Report {
  version: 1;
  language: "python" | "typescript";
  infermap_version: string;
  runner_version: string;
  ran_at: string;
  duration_seconds: number;
  scorecard: {
    overall: MetricSet;
    by_difficulty: Record<string, MetricSet>;
    by_category: Record<string, MetricSet>;
    by_tag: Record<string, MetricSet>;
  };
  per_case: Array<{
    id: string;
    f1: number;
    top1: number;
    mrr: number;
    expected_n: number;
    predicted_n: number;
    true_positives: number;
    false_positives: number;
    false_negatives: number;
    failed: boolean;
    failure_reason: string | null;
  }>;
  failed_cases: string[];
}

export interface BuildReportOptions {
  language: "python" | "typescript";
  infermapVersion: string;
  runnerVersion: string;
  durationSeconds: number;
}

const round6 = (n: number): number => Math.round(n * 1e6) / 1e6;

function metricSet(results: CaseResult[]): MetricSet {
  const counts = results.map((r) => ({ tp: r.tp, fp: r.fp, fn: r.fn }));
  const allPredictions: Prediction[] = [];
  for (const r of results) allPredictions.push(...r.predictions);
  return {
    f1: round6(microF1(counts)),
    top1: round6(macroMean(results.map((r) => r.top1))),
    mrr: round6(macroMean(results.map((r) => r.mrr))),
    ece: round6(expectedCalibrationError(allPredictions)),
    n: results.length,
  };
}

function bySingleKey(
  results: CaseResult[],
  key: (r: CaseResult) => string,
): Record<string, MetricSet> {
  const buckets = new Map<string, CaseResult[]>();
  for (const r of results) {
    const k = key(r);
    const arr = buckets.get(k) ?? [];
    arr.push(r);
    buckets.set(k, arr);
  }
  const sortedKeys = [...buckets.keys()].sort();
  const out: Record<string, MetricSet> = {};
  for (const k of sortedKeys) out[k] = metricSet(buckets.get(k)!);
  return out;
}

function byTag(results: CaseResult[]): Record<string, MetricSet> {
  const buckets = new Map<string, CaseResult[]>();
  for (const r of results) {
    for (const tag of r.tags) {
      const arr = buckets.get(tag) ?? [];
      arr.push(r);
      buckets.set(tag, arr);
    }
  }
  const sortedKeys = [...buckets.keys()].sort();
  const out: Record<string, MetricSet> = {};
  for (const k of sortedKeys) out[k] = metricSet(buckets.get(k)!);
  return out;
}

export function buildReport(results: CaseResult[], opts: BuildReportOptions): Report {
  return {
    version: REPORT_VERSION as 1,
    language: opts.language,
    infermap_version: opts.infermapVersion,
    runner_version: opts.runnerVersion,
    ran_at: new Date().toISOString(),
    duration_seconds: Math.round(opts.durationSeconds * 1e4) / 1e4,
    scorecard: {
      overall: metricSet(results),
      by_difficulty: bySingleKey(results, (r) => r.difficulty),
      by_category: bySingleKey(results, (r) => r.category),
      by_tag: byTag(results),
    },
    per_case: results.map((r) => ({
      id: r.caseId,
      f1: round6(r.f1),
      top1: round6(r.top1),
      mrr: round6(r.mrr),
      expected_n: r.tp + r.fn,
      predicted_n: r.tp + r.fp,
      true_positives: r.tp,
      false_positives: r.fp,
      false_negatives: r.fn,
      failed: r.failed,
      failure_reason: r.failureReason,
    })),
    failed_cases: results.filter((r) => r.failed).map((r) => r.caseId),
  };
}

export function validateReport(report: unknown, schemaPath: string): void {
  const schema = JSON.parse(readFileSync(schemaPath, "utf8"));
  const AjvCtor = (Ajv2020 as unknown as { default?: typeof Ajv2020 }).default ?? Ajv2020;
  const ajv = new (AjvCtor as any)({ allErrors: true });
  const addFormatsFn =
    (addFormats as unknown as { default?: typeof addFormats }).default ?? addFormats;
  (addFormatsFn as any)(ajv);
  const validate = ajv.compile(schema);
  if (!validate(report)) {
    const errors = (validate.errors ?? [])
      .map((e: any) => `${e.instancePath ?? ""} ${e.message ?? ""}`.trim())
      .join(", ");
    throw new Error(`Report failed schema validation: ${errors}`);
  }
}

export function writeReport(report: Report, outputPath: string, schemaPath: string): void {
  validateReport(report, schemaPath);
  writeFileSync(
    outputPath,
    JSON.stringify(report, null, 2) + "\n",
    "utf8",
  );
}
